src/java/org/apache/cassandra/db/compaction/BackgroundCompactionRunner.java

```diff
@@ -0,0 +1,396 @@
+    }
+
+    /**
+     * Marks a CFS for compaction. Since marked, it will become a possible candidate for compaction. The mark will be
+     * cleared when we actually run the compaction for CFS.
+     *
+     * @return a promise which will be completed when the mark is cleared
+     */
+    Future<RequestResult> requestCompaction(ColumnFamilyStore cfs)
+    {
+        logger.trace("Requested background compaction for {}.{}", cfs.getKeyspaceName(), cfs.getTableName());
+
+        if (checkExecutor.isShutdown())
+        {
+            logger.info("Executor has been shut down, background compactions check will not be scheduled");
+            return Futures.immediateCancelledFuture();
+        }
+
+        SettableFuture<RequestResult> futureResult = SettableFuture.create();
+        compactionRequests.compute(cfs, (ignored, current) -> {
+            if (current == null)
+                current = Collections.synchronizedList(new ArrayList<>());
+            current.add(futureResult);
+            return current;
+        });
+
+        if (!maybeScheduleNextCheck())
+        {
+            futureResult.cancel(false);
+        }
+        return futureResult;
+    }
+
+    void shutdown()
+    {
+        checkExecutor.shutdown();
+        compactionRequests.values().forEach(promises -> promises.forEach(p -> p.cancel(false)));
+    }
+
+    int getOngoingCompactionsCount(ColumnFamilyStore cfs)
+    {
+        return ongoingCompactions.count(cfs);
+    }
+
+    @Override
+    public void run()
+    {
+        logger.trace("Running background compactions check");
+
+        // We shuffle the CFSs for which the compaction was requested so that with each run we traverse those CFSs
+        // in different order and make each CFS have equal chance to be selected
+        ArrayList<ColumnFamilyStore> compactionRequestsList = new ArrayList<>(compactionRequests.keySet());
+        Collections.shuffle(compactionRequestsList, random);
+
+        for (ColumnFamilyStore cfs : compactionRequestsList)
+        {
+            List<SettableFuture<RequestResult>> promises = compactionRequests.remove(cfs);
+
+            logger.trace("Checking compaction request for {}.{} - there are {} awaiting promises",
+                         cfs.getKeyspaceName(),
+                         cfs.getTableName(),
+                         promises.size());
+
+            RequestResult completionStatus = canRunCompactionTasks(cfs, promises);
+            if (completionStatus != null)
+            {
+                promises.stream().filter(p -> !p.isDone()).forEach(p -> p.set(completionStatus));
+                continue;
+            }
+
+            logger.trace("Scheduling a background task check for {}.{} with {}",
+                         cfs.keyspace.getName(),
+                         cfs.name,
+                         cfs.getCompactionStrategy().getName());
+
+            Supplier<CompletableFuture<?>> compactionTask = Optional.ofNullable(lazyRunCompactionTasks(cfs))
```

Comment:
Nit: This creates some unnecessary garbage and is not that easy to see through. I would spell out 
```
if (compactionTask == null)
    compactionTask = lazyRunUpgradeTasks(cfs);
```

Why is the supplier necessary?

Reply:
I use the supplier because I suppose that I need to know whether there is anything to run or not before starting the tasks. If the returned supplier is `null` - there is nothing to run. Based on that, I can add the CFS to `ongoingCompactions` before actually starting the compaction tasks. I cannot add it after starting because it is a multiset and removal is tied to the completion of the task - so we would have a race.

I also didn't want to spread usages of `ongoingCompactions` between methods and have it modified only in a single place